app/services/scm/checkout_instructions_service.rb

```diff
@@ -0,0 +1,142 @@
+class Scm::CheckoutInstructionsService
+  attr_reader :repository, :user
+
+  def initialize(repository, user = User.current)
+    @repository = repository
+    @user = user
+  end
+
+  ##
+  # Retrieve the checkout URL using the repository vendor information
+  # It may additionally set a path parameter, if the repository supports subtree checkout
+  def checkout_url(path = nil)
+    repository.scm.checkout_url(repository, checkout_base_url, path)
+  end
+
+  ##
+  # Returns the checkout command from SCM adapter
+  # (e.g., `git clone`)
+  def checkout_command
+    repository.scm.checkout_command
+  end
+
+  ##
+  # Returns the checkout base URL as defined in settings.
+  def checkout_base_url
+    checkout_settings['base_url']
+  end
+
+  ##
+  # Returns the instructions defined in the settings.
+  def instructions
+    checkout_settings['text']
+  end
+
+  ##
+  # Returns true when the checkout URL may target a subtree of the repository.
+  def subtree_checkout?
+    repository.scm.subtree_checkout?
+  end
+
+  ##
+  # Returns whether the repository supports showing checkout information
+  # and has been configured for it.
+  def available?
+    checkout_enabled? &&
+      repository.supports_checkout_info? &&
+      checkout_base_url.present?
+  end
+
+  def checkout_enabled?
+    checkout_settings['enabled'].to_i > 0
+  end
+
+  ##
+  # Determines whether permissions for the given repository
+  # are available.
+  def permission?
+    repository.managed?
+  end
+
+  ##
+  # Returns one of the following permission symbols for the given user
+  #
+  # - :readwrite: When user is allowed to read and commit (:commit_access)
+  # - :read: When user is allowed to checkout the repository, but not commit (:browse_repository)
+  # - :none: Otherwise
+  #
+  # Note that this information is only applicable when the repository is managed,
+  # because otherwise OpenProject does not control the repository permissions.
+  # Use +permission?+ to check whether this is the case.
+  #
+  def permission
+    project = repository.project
+    if user.allowed_to?(:commit_access, project)
+      :readwrite
+    elsif user.allowed_to?(:browse_repository, project)
+      :read
+    else
+      :none
+    end
+  end
+
+  ##
+  # Returns whether the given user may checkout the repository
+  #
+  # Note that this information is only applicable when the repository is managed,
+  # because otherwise OpenProject does not control the repository permissions.
+  # Use +permission?+ to check whether this is the case.
+  def may_checkout?
+    [:readwrite, :read].include?(permission)
+  end
+
+  ##
+  # Returns whether the given user may commit to the repository
+  #
+  # Note that this information is only applicable when the repository is managed,
+  # because otherwise OpenProject does not control the repository permissions.
+  # Use +permission?+ to check whether this is the case.
+  def may_commit?
+    permission == :readwrite
```

Comment:
I somehow see `permission` returning an array of either `[]`, `[:read]` or `[:read, :write]`... I think you are not doing it, because the way you are doing it now is probably better suited to generate a human readable string...

I am not sure which way I would prefer...